### PR TITLE
Adopt pyright for typechecking

### PR DESCRIPTION
**Pull request by smackesey**
### Summary & Motivation

- Drop mypy configuration and BK steps
- Add [Pyright](https://github.com/microsoft/pyright) configuration and a single BK step that checks the whole repo. BK step uses `make pyright`, which can also be run locally.

Benefits. Pyright:

- Is 4-5x faster than mypy (it runs on NodeJS and not Python)
- Is deeply integrated with Pylance, the state-of-the-art Python language server for VSCode. This allows rapid incremental type-checking during development, which makes it easy to achieve parity between editor diagnostics and CI.
- Is much better maintained than mypy. Bugs are fixed fast.
- Has a more powerful type-checking engine than mypy, allowing more expressive types to be used.
- Can be used to assess type-completeness of our public API in addition to type-correctness of annotated code.

Limitations:

- There are a few cases where third-party libraries use mypy-specific extensions to make non-standard APIs legible to mypy. Most of these do not yet offer similar "adapter" functionality for pyright. However, as Python's typing capabilities grow, the need for third-party extensions is diminishing.

### How I Tested These Changes

BK

### CODE REVIEW

**Comment by smackesey** [comment]
Current dependencies on/for this PR:
* master
  * **PR #10900** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/10900" target="_blank">Graphite</a> 
    * **PR #10901** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/10901" target="_blank">Graphite</a> 
      * **PR #10907** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/10907" target="_blank">Graphite</a> 
        * **PR #10983** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/10983" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/10983?utm_source=stack-comment).

**Comment by schrockn** [comment]
I'm extremely excited to make this switch. but the approach here is simply too high risk and this PR is too large to be reviewable. We *have* to break this up into reviewable subcomponents. The only PRs that should be this large are purely mechanical (e.g. a rename of a commonly used utility function) to the point where they could be applied and replicated as a script.

This sequence of changes needs to end up in a state where, for example, your dagster-graphql changes are in 1 PR, the pendulum changes in another, and so forth.

We're chatting in Slack but wanted to make sure spectators know where this is at.

File: docs/dagit-screenshot/tox.ini
Comment (schrockn) [inline docs/dagit-screenshot/tox.ini:1]: is this deletion deliberate?

Comment (schrockn) [reply docs/dagit-screenshot/tox.ini:1]: this is only for mypy?

File: python_modules/dagster/dagster/_core/execution/job_backfill.py
Comment (schrockn) [inline python_modules/dagster/dagster/_core/execution/job_backfill.py:283]: this looks very problematic

File: python_modules/dagster/dagster/_core/host_representation/repository_location.py
Comment (schrockn) [inline python_modules/dagster/dagster/_core/host_representation/repository_location.py:452]: isn't it better to leave this untyped? why do this if you have the type: ignore below?

File: python_modules/dagster/dagster/_core/telemetry.py
Comment (schrockn) [inline python_modules/dagster/dagster/_core/telemetry.py:578]: looks problematic

File: python_modules/dagster/dagster/_grpc/impl.py
Comment (schrockn) [inline python_modules/dagster/dagster/_grpc/impl.py:275]: this looks scary. would prefer type: ignore for conservatism sake

File: scripts/check_schemas.py
Comment (schrockn) [inline scripts/check_schemas.py:56]: weird change to include but don't care

File: tox.ini
Comment (schrockn) [inline tox.ini:1]: This seems like a big change to our CI. Can you add a big comment block at the top explaining this?

**Review by schrockn — CHANGES_REQUESTED** [review]
This is very epic. I have a few comments that I'd like address, but this is essentially gtg.

File: tox.ini
Comment (smackesey) [reply tox.ini]: I changed the approach, no longer have a root-level tox

File: docs/dagit-screenshot/tox.ini
Comment (smackesey) [reply docs/dagit-screenshot/tox.ini]: Yes only `mypy` was run on it. It's an internal utility (glorified script) for taking dagit screenshots that doesn't have any unit tests.

File: scripts/run-pyright.py
Comment (schrockn) [inline scripts/run-pyright.py:242]: seems like this should be in a finally block

File: pyproject.toml
Comment (schrockn) [inline pyproject.toml:87]: wait so now we are turning this off?

File: pyright/alt-1/requirements.txt
Comment (schrockn) [inline pyright/alt-1/requirements.txt:11]: what is the plan for keeping this file up to date?

File: pyright/master/requirements.txt
Comment (schrockn) [inline pyright/master/requirements.txt:9]: Do we have a running list of known dependency conflicts? I'm sure that is in your head right now, and it would be useful to document.

File: pyright/alt-1/requirements.txt
Comment (schrockn) [inline pyright/alt-1/requirements.txt:80]: no references to internal in the OSS repo pleaase

**Review by schrockn — CHANGES_REQUESTED** [review]
req'ing changes based on the reference to internal.

I'm definitely a bit concerned about the maintainability of the pyright merging logic over time.

Does that provide any value beyond the reduction in CI jobs?

File: pyproject.toml
Comment (smackesey) [reply pyproject.toml:87]: In that DM I said we'd turn it off (but that handles only _some_ cases)-- I edited it maybe an hour or so after first sent:

>Given this, the best we can do is turn off reportPrivateImportUsage, which will affect pendulum and datadog but not e.g. sqlalchemy and several other libs that required import changes. (edited)

File: pyright/master/requirements.txt
Comment (smackesey) [reply pyright/master/requirements.txt:9]: I could figure it out with some digging but it's surprisingly hard to figure out what the actual conflicts are. The way I arrived at the current partition is by repeatedly `pip install --dry-run` on our whole package list and eliminating until I could get a successful install.

**Comment by smackesey** [comment]
> I'm definitely a bit concerned about the maintainability of the pyright merging logic over time.

I can see why you might be worried, but I'm not particularly concerned-- it's just some JSON massaging. The default pyright output is terrible on a lot of files and JSON output of `pyright` is not an afterthought-- the maintainer is always telling anyone who doesn't like it to "use JSON output and write a script".

> Does that provide any value beyond the reduction in CI jobs?

The advantage is that it lets you run one command and see combined output from all pyright runs. Without some kind of similar merge post-processing, you'd have to scroll up a bunch in the terminal to see the output from successive runs.

File: pyright/alt-1/requirements.txt
Comment (smackesey) [reply pyright/alt-1/requirements.txt]: Yeah this was there as a placeholder I intend to remove before merge (they're gone now)-- I was looking for a solution for `quickstart-*` examples, since they depend on `dagster-cloud`.

But having looked at the code of these, they don't actually import from `dagster_cloud` (or most of their deps), so it looks like they can successfully be type-checked in the master env without needing to install their deps.

File: pyright/alt-1/requirements.txt
Comment (smackesey) [reply pyright/alt-1/requirements.txt:11]: There is no special plan-- I think it can more-or-less be treated like any one of our gazillion toxfiles, which also have deps that have to be kept in sync with corresponding setup.py files.

I would love it if we could get everything in one env or there were some kind of more automated system for partitioning our packages into envs, but I'm not aware of a better option.

File: scripts/run-pyright.py
Comment (smackesey) [reply scripts/run-pyright.py]: you're right

File: pyproject.toml
Comment (schrockn) [reply pyproject.toml:87]: Ah got thanks for clarification

File: pyright/master/requirements.txt
Comment (schrockn) [reply pyright/master/requirements.txt:9]: Got it. Not asking you to dig just to document what is resident in your brain cache.

**Review by schrockn — APPROVED** [review]
Ok let's do this then.

I'd like @jmsanders or @alangenfeld  to give final blessing on the buildkite changes before you merge this sucker.

An exciting step forward for our codebase! Thank you for doing this I know it was an extraordinary amount of work.

File: .buildkite/dagster-buildkite/dagster_buildkite/package_spec.py
Comment (jmsanders) [inline .buildkite/dagster-buildkite/dagster_buildkite/package_spec.py:204]: What are we trying to guard against here?

**Review by jmsanders — APPROVED** [review]

File: .buildkite/dagster-buildkite/dagster_buildkite/package_spec.py
Comment (smackesey) [reply .buildkite/dagster-buildkite/dagster_buildkite/package_spec.py:204]: We return a `CommandStep` instead of a `GroupStep` if there's only one step, to avoid having to unnecessarily click through a group layer in BK.

**Pull request merged**
